aspnetcore/fundamentals/http-requests.md

```diff
@@ -0,0 +1,394 @@
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddHttpClient<GitHubService>(c =>
+    {
+        c.BaseAddress = new Uri("https://api.github.com/");
+
+        c.DefaultRequestHeaders.Add("Accept", "application/vnd.github.v3+json"); // GitHub API versioning
+        c.DefaultRequestHeaders.Add("User-Agent", "HttpClientFactory-Sample"); // GitHub requires a user-agent
+    });
+    services.AddMvc();
+}
+```
+
+It's possible to entirely encapsulate the `HttpClient` within a typed client. Rather than exposing it as a property, public methods can be provided which call the `HttpClient` instance internally.
+
+```csharp
+public class ValuesService
+{
+    private readonly HttpClient _httpClient;
+    private readonly IMemoryCache _cache;
+    private readonly ILogger<ValuesService> _logger;
+
+    public ValuesService() { }
+
+    public ValuesService(HttpClient client, IMemoryCache cache, ILogger<ValuesService> logger)
+    {
+        _httpClient = client;
+        _cache = cache;
+        _logger = logger;
+    }
+
+    public async Task<IEnumerable<string>> GetValues()
+    {
+        var result = await _httpClient.GetAsync("api/values");
+        var resultObj = Enumerable.Empty<string>();
+
+        if (result.IsSuccessStatusCode)
+        {
+            resultObj = JsonConvert.DeserializeObject<IEnumerable<string>>(await result.Content.ReadAsStringAsync());
+            _cache.Set("GetValue", resultObj);
+        }
+        else
+        {
+            if (_cache.TryGetValue("GetValue", out resultObj))
+            {
+                _logger.LogWarning("Returning cached values as the values service is unavailable.");
+                return resultObj;
+            }
+            result.EnsureSuccessStatusCode();
+        }        
+        return resultObj;
+    }
+}
+```
+
+In the preceding code, the `HttpClient` is stored as a private field. All access to make external calls goes through the `GetValues` method.
+
+## Generated clients
+
+`HttpClientFactory` can be used in combination with other third-party libraries such as [Refit](https://github.com/paulcbetts/refit). Refit is a REST library for .NET which turns REST APIs into live interfaces. An implementation of the interface is generated dynamically by the `RestService`, using `HttpClient` to make the external HTTP calls.
+
+An interface and a reply are defined to represent the external API and its response:
+
+```csharp
+public interface IHelloClient
+{
+    [Get("/helloworld")]
+    Task<Reply> GetMessageAsync();
+}
+
+public class Reply
+{
+    public string Message { get; set; }
+}
+```
+
+A typed client can be added, using Refit to generate the implementation:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddHttpClient("hello", c =>
+    {
+        c.BaseAddress = new Uri("http://localhost:5000");
+    })
+    .AddTypedClient(c => Refit.RestService.For<IHelloClient>(c));
```

Comment:
We also have another variant here that will takes `<TInterface, TImpl>`.

One other comment I have about this section is that this style of configuration is in no way specific to refit. It can work with any type that accepts an `HttpClient`, or you can provide your own factory methods.